File: 23game.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <cmath>

using namespace std;

// Función para verificar si un número es primo
bool es_primo(int num) {
    if (num < 2)
        return false;
    for (int i = 2; i <= sqrt(num); i++) {
        if (num % i == 0)
            return false;
    }
    return true;
}

// Función para lanzar dados
vector<int> lanzar_dados(int num_dados = 2) {
    vector<int> dados;
    for (int i = 0; i < num_dados; i++) {
        dados.push_back(rand() % 6 + 1);  // Número aleatorio entre 1 y 6
    }
    return dados;
}

// Función para jugar una ronda
int jugar_ronda() {
    int total = 0;

    // Lanzar dos dados inicialmente
    vector<int> dados = lanzar_dados(2);
    cout << "Lanzamiento inicial: [" << dados[0] << ", " << dados[1] << "]" << endl;

    // Sumar los valores de los dados
    total = dados[0] + dados[1];

    // Verificar si el jugador sacó dobles
    if (dados[0] == dados[1]) {
        cout << "¡Sacaste dobles! Lanzarás 3 dados adicionales." << endl;
        vector<int> dados_extra = lanzar_dados(3);
        cout << "Lanzamiento adicional con 3 dados: [" << dados_extra[0] << ", " << dados_extra[1] << ", " << dados_extra[2] << "]" << endl;
        total = dados_extra[0] + dados_extra[1] + dados_extra[2];
    }

    // Verificar si sacó 12
    if (total == 12) {
        char decision;
        cout << "Sacaste 12. ¿Deseas lanzar solo 1 dado adicional (S/N)? ";
        cin >> decision;
        if (decision == 'S' || decision == 's') {
            int dado_extra = lanzar_dados(1)[0];
            cout << "Lanzaste 1 dado adicional: " << dado_extra << endl;
            total += dado_extra;
        }
    }

    cout << "Total hasta ahora: " << total << endl;
    return total;
}

// Función para determinar el ganador de una ronda
int determinar_ganador(int total_jugador1, int total_jugador2) {
    // Comprobar si los números son primos
    bool primo_jugador1 = es_primo(total_jugador1);
    bool primo_jugador2 = es_primo(total_jugador2);

    if (primo_jugador1 && !primo_jugador2) {
        cout << "El jugador 1 gana con " << total_jugador1 << ", que es un número primo." << endl;
        return 1;
    } else if (primo_jugador2 && !primo_jugador1) {
        cout << "El jugador 2 gana con " << total_jugador2 << ", que es un número primo." << endl;
        return 2;
    } else if (primo_jugador1 && primo_jugador2) {
        if (total_jugador1 > total_jugador2) {
            cout << "El jugador 1 gana con " << total_jugador1 << ", que es un número primo mayor." << endl;
            return 1;
        } else {
            cout << "El jugador 2 gana con " << total_jugador2 << ", que es un número primo mayor." << endl;
            return 2;
        }
    } else {
        if (total_jugador1 <= 23 && total_jugador2 <= 23) {
            if (total_jugador1 > total_jugador2) {
                cout << "El jugador 1 gana con " << total_jugador1 << ", más cercano a 23." << endl;
                return 1;
            } else {
                cout << "El jugador 2 gana con " << total_jugador2 << ", más cercano a 23." << endl;
                return 2;
            }
        } else if (total_jugador1 <= 23) {
            cout << "El jugador 1 gana con " << total_jugador1 << ", más cercano a 23." << endl;
            return 1;
        } else if (total_jugador2 <= 23) {
            cout << "El jugador 2 gana con " << total_jugador2 << ", más cercano a 23." << endl;
            return 2;
        } else {
            cout << "Ambos jugadores superaron 23. No hay ganador." << endl;
            return 0;
        }
    }
}

// Función principal para jugar Prime Jack al mejor de 5
void jugar_prime_jack() {
    cout << "¡Bienvenidos a Prime Jack al mejor de 5!" << endl;

    int rondas_jugadas = 0;
    int puntos_jugador1 = 0;
    int puntos_jugador2 = 0;

    while (puntos_jugador1 < 3 && puntos_jugador2 < 3) {
        rondas_jugadas++;
        cout << "\n--- Ronda " << rondas_jugadas << " ---" << endl;

        cout << "\nJugador 1:" << endl;
        int total_jugador1 = jugar_ronda();

        cout << "\nJugador 2:" << endl;
        int total_jugador2 = jugar_ronda();

        // Determinar el ganador de la ronda
        int ganador = determinar_ganador(total_jugador1, total_jugador2);
        if (ganador == 1) {
            puntos_jugador1++;
            cout << "El Jugador 1 gana la ronda " << rondas_jugadas << "." << endl;
        } else if (ganador == 2) {
            puntos_jugador2++;
            cout << "El Jugador 2 gana la ronda " << rondas_jugadas << "." << endl;
        } else {
            cout << "No hay ganador en la ronda " << rondas_jugadas << "." << endl;
        }

        // Mostrar los puntos acumulados
        cout << "Puntos Jugador 1: " << puntos_jugador1 << endl;
        cout << "Puntos Jugador 2: " << puntos_jugador2 << endl;
    }

    // Declarar el ganador final
    if (puntos_jugador1 == 3) {
        cout << "\n¡El Jugador 1 ha ganado el juego al mejor de 5!" << endl;
    } else {
        cout << "\n¡El Jugador 2 ha ganado el juego al mejor de 5!" << endl;
    }
}

int main() {
    srand(time(0));  // Inicializar la semilla para números aleatorios
    jugar_prime_jack();
    return 0;
}

File: 23game.py
import random

# Función para verificar si un número es primo
def es_primo(num):
    if num < 2:
        return False
    for i in range(2, int(num ** 0.5) + 1):
        if num % i == 0:
            return False
    return True

# Función para lanzar dados
def lanzar_dados(num_dados=2):
    return [random.randint(1, 6) for _ in range(num_dados)]

# Función para jugar una ronda
def jugar_ronda():
    total = 0
    
    # Lanzar dos dados inicialmente
    dados = lanzar_dados(2)
    print(f"Lanzamiento inicial: {dados}")

    # Sumar los valores de los dados
    total = sum(dados)

    # Verificar si el jugador sacó dobles
    if dados[0] == dados[1]:
        print("¡Sacaste dobles! Lanzarás 3 dados adicionales.")
        dados_extra = lanzar_dados(3)
        print(f"Lanzamiento adicional con 3 dados: {dados_extra}")
        total = sum(dados_extra)
    
    # Verificar si sacó 12
    if total == 12:
        decision = input("Sacaste 12. ¿Deseas lanzar solo 1 dado adicional (S/N)? ").strip().lower()
        if decision == 's':
            dado_extra = lanzar_dados(1)[0]
            print(f"Lanzaste 1 dado adicional: {dado_extra}")
            total += dado_extra

    print(f"Total hasta ahora: {total}")
    return total

# Función para determinar el ganador de una ronda
def determinar_ganador(total_jugador1, total_jugador2):
    # Comprobar si los números son primos
    primo_jugador1 = es_primo(total_jugador1)
    primo_jugador2 = es_primo(total_jugador2)
    
    if primo_jugador1 and not primo_jugador2:
        print(f"El jugador 1 gana con {total_jugador1}, que es un número primo.")
        return 1
    elif primo_jugador2 and not primo_jugador1:
        print(f"El jugador 2 gana con {total_jugador2}, que es un número primo.")
        return 2
    elif primo_jugador1 and primo_jugador2:
        if total_jugador1 > total_jugador2:
            print(f"El jugador 1 gana con {total_jugador1}, que es un número primo mayor.")
            return 1
        else:
            print(f"El jugador 2 gana con {total_jugador2}, que es un número primo mayor.")
            return 2
    else:
        if total_jugador1 <= 23 and total_jugador2 <= 23:
            if total_jugador1 > total_jugador2:
                print(f"El jugador 1 gana con {total_jugador1}, más cercano a 23.")
                return 1
            else:
                print(f"El jugador 2 gana con {total_jugador2}, más cercano a 23.")
                return 2
        elif total_jugador1 <= 23:
            print(f"El jugador 1 gana con {total_jugador1}, más cercano a 23.")
            return 1
        elif total_jugador2 <= 23:
            print(f"El jugador 2 gana con {total_jugador2}, más cercano a 23.")
            return 2
        else:
            print("Ambos jugadores superaron 23. No hay ganador.")
            return 0

# Función principal para jugar Prime Jack al mejor de 5
def jugar_prime_jack():
    print("¡Bienvenidos a Prime Jack al mejor de 5!")

    rondas_jugadas = 0
    puntos_jugador1 = 0
    puntos_jugador2 = 0

    while puntos_jugador1 < 3 and puntos_jugador2 < 3:
        rondas_jugadas += 1
        print(f"\n--- Ronda {rondas_jugadas} ---")

        print("\nJugador 1:")
        total_jugador1 = jugar_ronda()

        print("\nJugador 2:")
        total_jugador2 = jugar_ronda()

        # Determinar el ganador de la ronda
        ganador = determinar_ganador(total_jugador1, total_jugador2)
        if ganador == 1:
            puntos_jugador1 += 1
            print(f"El Jugador 1 gana la ronda {rondas_jugadas}.")
        elif ganador == 2:
            puntos_jugador2 += 1
            print(f"El Jugador 2 gana la ronda {rondas_jugadas}.")
        else:
            print(f"No hay ganador en la ronda {rondas_jugadas}.")

        # Mostrar los puntos acumulados
        print(f"Puntos Jugador 1: {puntos_jugador1}")
        print(f"Puntos Jugador 2: {puntos_jugador2}")

    # Declarar el ganador final
    if puntos_jugador1 == 3:
        print("\n¡El Jugador 1 ha ganado el juego al mejor de 5!")
    else:
        print("\n¡El Jugador 2 ha ganado el juego al mejor de 5!")

# Ejecutar el juego
jugar_prime_jack()